### PR TITLE
Stop Loss, Take Profit, Rug checking and Price checking implementation

### PR DESCRIPTION
**Pull request by oboshto**
- feature: rug checking
- feature: price checking via birdeye
- feature: stop loss
- feature: take profit
- remove shitty logging emojis

### CODE REVIEW

**Comment by Hid3away** [comment]
Is this safe?

**Comment by oboshto** [comment]
@Hid3away depends on your strategy

**Comment by virtuman** [comment]
@fdundjer - I refactored code, added DB support and split logic, maybe i can screen share, to show you the code. And if you like, i'll open a pull request, or if you don't like the style, i'll just continue in my own branch. I added a bit of clarity and extra "module"-based logic as well as rug checkers, solsniffer, rugcheck, etc. optimized and added  price tracking,  retries, wherever make sense, etc. 

i feel that opening pull-request at this point will just turn it into a never-ending story...

**Comment by davidxiaodev** [comment]
> @Hid3away

thanks, sounds great,
but one question, wonder if we can get live transaction events and parse price information from there?
regular fetching price seems not good way, i think for sure it will miss some price info

it's a question, i am new to Solana, so don't know if it's possible?

Thanks

**Comment by BartBGSAV** [comment]
Tested this out. I can not see the TP or SL working :)
Would need some further explanation

**Comment by oboshto** [comment]
@BartBGSAV did you grab latest commit https://github.com/snipe-it-com/solana-sniper-bot/pull/82/commits/4adb22fa7f1b5cc02f0d2c40cd6652458b55e4a2? I fixed that behaviour yesterday

**Comment by oboshto** [comment]
@VasilisTako do you mean to use some of solana oracles? or could you please point to the direction

**Comment by Journeytojah** [comment]
@oboshto are you on discord?

**Comment by oboshto** [comment]
@Journeytojah savadikap https://discordapp.com/users/600250420480442390

**Comment by BartBGSAV** [comment]
> @BartBGSAV did you grab latest commit [4adb22f](https://github.com/snipe-it-com/solana-sniper-bot/commit/4adb22fa7f1b5cc02f0d2c40cd6652458b55e4a2)? I fixed that behaviour yesterday

yes sir.
I am on the discord ( BG ) you can pm me. It fetched the price at the buy but afterwards it didnt. i can output a log

**Comment by oboshto** [comment]
@BartBGSAV i can't find you on discord. pm me please (https://github.com/snipe-it-com/solana-sniper-bot/pull/82#issuecomment-2056660726)

**Comment by Hid3away** [comment]
Could you add me also please ? @VespertineD2K

**Comment by fdundjer** [comment]
@oboshto Great work. Clean and well documented. Reach out to me on discord to give you the role.

File: buy.ts
[the remark this answers is not in the document]
Comment (oboshto) [reply buy.ts:1]: just activate free package here (it's limited) but you will have the access to this endpoint (idk about the future but works for now)

[the remark this answers is not in the document]
Comment (fpena06) [reply buy.ts:1]: I have paid Birdeye and experiencing some token prices not found. I checked with Postman and it's not an issue with the bot.

[the remark this answers is not in the document]
Comment (oboshto) [reply buy.ts:1]: @fpena06 yes, you have to wait around 1-5 sec before fetch birdeye's endpoint (you may adjust RUG_CHECK_WAIT_TIMEOUT_SECONDS in .env). but i will fix that behaviour, it should get data on chain

**Comment by fpena06** [comment]
> @oboshto Great work. Clean and well documented. Reach out to me on discord to give you the role.

Hi, is there a discord for the bot? Thanks

**Comment by Journeytojah** [comment]
> > @oboshto Great work. Clean and well documented. Reach out to me on discord to give you the role.
> 
> Hi, is there a discord for the bot? Thanks

https://discord.gg/uTKwDcwy

**Comment by Klaudioz** [comment]
> > > @oboshto Great work. Clean and well documented. Reach out to me on discord to give you the role.
> > 
> > 
> > Hi, is there a discord for the bot? Thanks
> 
> https://discord.gg/uTKwDcwy

I got this clicking the link:

image

**Pull request closed without merging**